### ISSUE TITLE
9 Anti-Patterns Every Programmer Should Be Aware Of

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#144]
9 Anti-Patterns Every Programmer Should Be Aware Of<br>
<br>
https://ift.tt/1H2KqIE<br>
<br>
Sahand Saba<br>
<br>
<p>A healthy dose of self-criticism is fundamental to professional and personal growth. When it comes to programming, this sense of self-criticism requires the ability to detect unproductive or counter-productive patterns in designs, code, processes, and behaviour. This is why a knowledge of anti-patterns is very useful for any programmer. This article is a discussion of anti-patterns that I have found to be recurring, ordered roughly based on how often I have come across them, and how long it took to undo the damage they caused.</p>
<p>Some of the anti-patterns discussed have elements in common with cognitive biases, or are directly caused by them. Links to relevant cognitive biases are provided as we go along in the article. Wikipedia also has a nice <a href="http://en.wikipedia.org/wiki/List_of_cognitive_biases">list of cognitive biases</a> for your reference.</p>
<p>And before starting, let's remember that dogmatic thinking stunts growth and innovation so consider the list as a set of guidelines and not written-in-stone rules. And if I missed anything that you consider to be important, feel free to comment below!</p>
<div>
<p>List of anti-patterns covered:</p>
<ul>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#premature-optimization">1&nbsp;&nbsp;&nbsp;Premature Optimization</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#bikeshedding">2&nbsp;&nbsp;&nbsp;Bikeshedding</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#analysis-paralysis">3&nbsp;&nbsp;&nbsp;Analysis Paralysis</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#god-class">4&nbsp;&nbsp;&nbsp;God Class</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#fear-of-adding-classes">5&nbsp;&nbsp;&nbsp;Fear of Adding Classes</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#inner-platform-effect">6&nbsp;&nbsp;&nbsp;Inner-platform Effect</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#magic-numbers-and-strings">7&nbsp;&nbsp;&nbsp;Magic Numbers and Strings</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#management-by-numbers">8&nbsp;&nbsp;&nbsp;Management by Numbers</a></li>
<li><a href="http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#useless-poltergeist-classes">9&nbsp;&nbsp;&nbsp;Useless (Poltergeist) Classes</a></li>
</ul>
</div>
<div>
<h2>1&nbsp;&nbsp;&nbsp;Premature Optimization</h2>
<blockquote>
<span>We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%.</span> <span>Donald Knuth</span>
</blockquote>
<blockquote>
<span>Although never is often better than *right* now.</span> <span>Tim Peters, The Zen of Python</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Optimizing before you have enough information to make educated conclusions about where and how to do the optimization.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>It is very difficult to know exactly what will be the bottleneck in practice. Attempting to optimize prior to having empirical data is likely to end up increasing code complexity and room for bugs with negligible improvements.</p>
</div>
<div>
<h3>How to avoid it</h3>
<p>Prioritize writing clean and readable code that works first, using known and tested algorithms and tools. Use profiling tools when needed to find bottlenecks and optimize the priorities. Rely on measurements and not guesses and speculation.</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>Caching before profiling to find the bottlenecks. Using complicated and unproven "heuristics" instead of a known mathematically correct algorithm. Choosing a new and untested experimental web framework that can theoretically reduce request latency under heavy loads while you are in early stages and your servers are idle most of the time.</p>
</div>
<div>
<h3>The tricky part</h3>
<p>The tricky part is knowing when the optimization is premature. It's important to plan in advance for growth. Choosing designs and platforms that will allow for easy optimization and growth is key here. It's also possible to use "premature optimization" as an excuse to justify writing bad code. Example: writing an <span>\(O(n^2)\)</span> algorithm to solve a problem when a simpler, mathematically correct, <span>\(O(n)\)</span> algorithm exists, simply because the simpler algorithm is harder to understand.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Profile before optimizing. Avoid trading simplicity for efficiency until it is needed, backed by empirical evidence.</p>
</div>
</div>
<div>
<h2>2&nbsp;&nbsp;&nbsp;Bikeshedding</h2>
<blockquote>
<span>Every once in a while we'd interrupt that to discuss the typography and the color of the cover. And after each discussion, we were asked to vote. I thought it would be most efficient to vote for the same color we had decided on in the meeting before, but it turned out I was always in the minority! We finally chose red. (It came out blue.)</span> <span>Richard Feynman, What Do You Care What Other People Think?</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Tendency to spend excessive amounts of time debating and deciding on trivial and often subjective issues.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>It's a waste of time. Poul-Henning Kamp goes into depth in an <a href="http://phk.freebsd.dk/sagas/bikeshed.html">excellent email here</a>.</p>
</div>
<div>
<h3>How to avoid it</h3>
<p>Encourage team members to be aware of this tendency, and to prioritize reaching a decision (vote, flip a coin, etc. if you have to) when you notice it. Consider A/B testing later to revisit the decision, when it is meaningful to do so (e.g. deciding between two different UI designs), instead of further internal debating.</p>
<div>
<img alt="Richard Feynman teaching a lecture." src="http://sahandsaba.com/images/feynman.jpg">
<p>Richard Feynman was not a fan of bikeshedding.</p>
</div>
</div>
<div>
<h3>Examples and signs</h3>
<p>Spending hours or days debating over what background color to use in your app, or whether to put a button on the left or the right of the UI, or to use tabs instead of spaces for indentation in your code base.</p>
</div>
<div>
<h3>The tricky part</h3>
<p>Bikeshedding is easier to notice and prevent in my opinion than premature optimization. Just try to be aware of the amount of time spent on making a decision and contrast that with how trivial the issue is, and intervene if necessary.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Avoid spending too much time on trivial decisions.</p>
</div>
</div>
<div>
<h2>3&nbsp;&nbsp;&nbsp;Analysis Paralysis</h2>
<blockquote>
<span>Want of foresight, unwillingness to act when action would be simple and effective, lack of clear thinking, confusion of counsel [...] these are the features which constitute the endless repetition of history.</span> <span>Winston Churchill, Parliamentary Debates</span>
</blockquote>
<blockquote>
<span>Now is better than never.</span> <span>Tim Peters, The Zen of Python</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Over-analyzing to the point that it prevents action and progress.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>Over-analyzing can slow down or stop progress entirely. In the extreme cases, the results of the analysis can become obsolete by the time they are done, or worse, the project might never leave the analysis phase. It is also easy to assume that more information will help decisions when the decision is a difficult one to make &#8213; see <a href="http://en.wikipedia.org/wiki/Information_bias_(psychology)">information bias</a> and <a href="http://en.wikipedia.org/wiki/Illusion_of_validity">validity bias</a>.</p>
</div>
<div>
<h3>How to avoid it</h3>
<p>Again, awareness helps. Emphasize iterations and improvements. Each iteration will provide more feedback with more data points that can be used for more meaningful analysis. Without the new data points, more analysis will become more and more speculative.</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>Spending months or even years deciding on a project's requirements, a new UI, or a database design.</p>
</div>
<div>
<h3>The tricky part</h3>
<p>It can be tricky to know when to move from planning, requirement gathering and design, to implementation and testing.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Prefer iterating to over-analyzing and speculation.</p>
</div>
</div>
<div>
<h2>4&nbsp;&nbsp;&nbsp;God Class</h2>
<blockquote>
<span>Simple is better than complex.</span> <span>Tim Peters, The Zen of Python</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Classes that control many other classes and have many dependencies and lots of responsibilities.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>God classes tend to grow to the point of becoming maintenance nightmares &#8213; because they violate the single-responsibility principle, they are hard to unit-test, debug, and document.</p>
</div>
<div>
<h3>How to avoid it</h3>
<p>Avoid having classes turn into God classes by breaking up the responsibilities into smaller classes with a single clearly-defined, unit-tested, and documented responsibility. Also see "Fear of Adding Classes" below.</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>Look for class names containing "manager", "controller", "driver", "system", or "engine". Be suspicious of classes that import or depend on many other classes, control too many other classes, or have many methods performing unrelated tasks.</p>
<div>
<p>God classes know about too many classes and/or control too many.</p>
</div>
</div>
<div>
<h3>The tricky part</h3>
<p>As projects age and requirements and the number of engineers grow, small and well-intentioned classes turn into God classes slowly. Refactoring such classes can become a significant task.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Avoid large classes with too many responsibilities and dependencies.</p>
</div>
</div>
<div>
<h2>5&nbsp;&nbsp;&nbsp;Fear of Adding Classes</h2>
<blockquote>
<span>Sparse is better than dense.</span> <span>Tim Peters, The Zen of Python</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Belief that more classes necessarily make designs more complicated, leading to a fear of adding more classes or breaking large classes into several smaller classes.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>Adding classes can help reduce complexity significantly. Picture a big tangled ball of yarns. When untangled, you will have several separated yarns instead. Similarly, several simple, easy-to-maintain and easy-to-document classes are much preferable to a single large and complex class with many responsibilities (see the God Class anti-pattern above).</p>
<div>
<img alt="Tangled ball of yarn" src="http://sahandsaba.com/images/yarn.jpg">
<p>A tangled ball of yarn. Large classes have a tendency to turn into the software equivalent of this. (<a href="https://www.flickr.com/photos/absolut_feli/7162673224">Photo by absolut_feli on Flickr</a>)</p>
</div>
</div>
<div>
<h3>How to avoid it</h3>
<p>Be aware of when additional classes can simplify the design and decouple unnecessarily coupled parts of your code.</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>As an easy example consider the following:</p>
<div>
<pre>
<span>class</span> <span>Shape</span><span>:</span>
    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>,</span> <span>shape_type</span><span>,</span> <span>*</span><span>args</span><span>):</span>
        <span>self</span><span>.</span><span>shape_type</span> <span>=</span> <span>shape_type</span>
        <span>self</span><span>.</span><span>args</span> <span>=</span> <span>args</span>

    <span>def</span> <span>draw</span><span>(</span><span>self</span><span>):</span>
        <span>if</span> <span>self</span><span>.</span><span>shape_type</span> <span>==</span> <span>"circle"</span><span>:</span>
            <span>center</span> <span>=</span> <span>self</span><span>.</span><span>args</span><span>[</span><span>0</span><span>]</span>
            <span>radius</span> <span>=</span> <span>self</span><span>.</span><span>args</span><span>[</span><span>1</span><span>]</span>
            <span># Draw a circle...</span>
        <span>elif</span> <span>self</span><span>.</span><span>shape_type</span> <span>==</span> <span>"rectangle"</span><span>:</span>
            <span>pos</span> <span>=</span> <span>self</span><span>.</span><span>args</span><span>[</span><span>0</span><span>]</span>
            <span>width</span> <span>=</span> <span>self</span><span>.</span><span>args</span><span>[</span><span>1</span><span>]</span>
            <span>height</span> <span>=</span> <span>self</span><span>.</span><span>args</span><span>[</span><span>2</span><span>]</span>
            <span># Draw rectangle...</span>
</pre>
</div>
<p>Now compare it with the following:</p>
<div>
<pre>
<span>class</span> <span>Shape</span><span>:</span>
    <span>def</span> <span>draw</span><span>(</span><span>self</span><span>):</span>
        <span>raise</span> <span>NotImplemented</span><span>(</span><span>"Subclasses of Shape should implement method 'draw'."</span><span>)</span>

<span>class</span> <span>Circle</span><span>(</span><span>Shape</span><span>):</span>
    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>,</span> <span>center</span><span>,</span> <span>radius</span><span>):</span>
        <span>self</span><span>.</span><span>center</span> <span>=</span> <span>center</span>
        <span>self</span><span>.</span><span>radius</span> <span>=</span> <span>radius</span>

    <span>def</span> <span>draw</span><span>(</span><span>self</span><span>):</span>
        <span># Draw a circle...</span>

<span>class</span> <span>Rectangle</span><span>(</span><span>Shape</span><span>):</span>
    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>,</span> <span>pos</span><span>,</span> <span>width</span><span>,</span> <span>height</span><span>):</span>
        <span>self</span><span>.</span><span>pos</span> <span>=</span> <span>pos</span>
        <span>self</span><span>.</span><span>width</span> <span>=</span> <span>width</span>
        <span>self</span><span>.</span><span>height</span> <span>=</span> <span>height</span>

    <span>def</span> <span>draw</span><span>(</span><span>self</span><span>):</span>
        <span># Draw a rectangle...</span>
</pre>
</div>
<p>Of course, this is an obvious example, but it illustrates the point: larger classes with conditional or complicated logic in them can, and often should, be broken down into simpler classes. The resulting code will have more classes but will be simpler.</p>
</div>
<div>
<h3>The tricky part</h3>
<p>Adding classes is not a magic bullet. Simplifying the design by breaking up large classes requires thoughtful analysis of the responsibilities and requirements.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>More classes are not necessarily a sign of bad design.</p>
</div>
</div>
<div>
<h2>6&nbsp;&nbsp;&nbsp;Inner-platform Effect</h2>
<blockquote>
<span>Those who do not understand Unix are condemned to reinvent it, poorly.</span> <span>Henry Spencer</span>
</blockquote>
<blockquote>
<span>Any sufficiently complicated C or Fortran program contains an ad hoc, informally-specified, bug-ridden, slow implementation of half of Common Lisp.</span> <span>Greenspun's tenth rule</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>The tendency for complex software systems to re-implement features of the platform they run in or the programming language they are implemented in, usually poorly.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>Platform-level tasks such as job scheduling and disk cache buffers are not easy to get right. Poorly designed solutions are prone to introduce bottlenecks and bugs, especially as the system scales up. And recreating alternative language constructs to achieve what is already possible in the language leads to difficult to read code and a steeper learning curve for anyone new to the code base. It can also limit the usefulness of refactoring and code analysis tools.</p>
</div>
<div>
<h3>How to avoid it</h3>
<p>Learn to use the platform or features provided by your OS or platform instead. Avoid the temptation to create language constructs that rival existing constructs (especially if it's because you are not used to a new language and miss your old language's features).</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>Using your MySQL database as a job queue. Reimplementing your own disk buffer cache mechanism instead of relying on your OS's. Writing a task scheduler for your web-server in PHP. Defining macros in C to allow for Python-like language constructs.</p>
</div>
<div>
<h3>The tricky part</h3>
<p>In very rare cases, it might be necessary re-implement parts of the platform (JVM, Firefox, Chrome, etc.).</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Avoid re-inventing what your OS or development platform already does well.</p>
</div>
</div>
<div>
<h2>7&nbsp;&nbsp;&nbsp;Magic Numbers and Strings</h2>
<blockquote>
<span>Explicit is better than implicit.</span> <span>Tim Peters, The Zen of Python</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Using unnamed numbers or string literals instead of named constants in code.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>The main problem is that the semantics of the number or string literal is partially or completely hidden without a descriptive name or another form of annotation. This makes understanding the code harder, and if it becomes necessary to change the constant, search and replace or other refactoring tools can introduce subtle bugs. Consider the following piece of code:</p>
<div>
<pre>
<span>def</span> <span>create_main_window</span><span>():</span>
    <span>window</span> <span>=</span> <span>Window</span><span>(</span><span>600</span><span>,</span> <span>600</span><span>)</span>
    <span># etc...</span>
</pre>
</div>
<p>What are the two numbers there? Assume the first is window width and the second in window height. If it ever becomes necessary to change the width to 800 instead, a search and replace would be dangerous since it would change the height in this case too, and perhaps other occurrences of the number 600 in the code base.</p>
<p>String literals might seem less prone to these issues but having unnamed string literals in code makes internationalization harder, and can introduce similar issues to do with instances of the same literal having different semantics. For example, homonyms in English can cause a similar issue with search and replace; consider two occurrences of "point", one in which it refers to a noun (as in "she has a point") and the other as a verb (as in "to point out the differences..."). Replacing such string literals with a string retrieval mechanism that allows you to clearly indicate the semantics can help distinguish these two cases, and will also come in handy when you send the strings for translation.</p>
</div>
<div>
<h3>How to avoid it</h3>
<p>Use named constants, resource retrieval methods, or annotations.</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>Simple example is shown above. This particular anti-pattern is very easy to detect (except for a few tricky cases mentioned below.)</p>
</div>
<div>
<h3>The tricky part</h3>
<p>There is a narrow grey area where it can be hard to tell if certain numbers are magic numbers or not. For example the number 0 for languages with zero-based indexing. Other examples are use of 100 to calculate percentages, 2 to check for parity, etc.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Avoid having unexplained and unnamed numbers and string literals in code.</p>
</div>
</div>
<div>
<h2>8&nbsp;&nbsp;&nbsp;Management by Numbers</h2>
<blockquote>
<span>Measuring programming progress by lines of code is like measuring aircraft building progress by weight.</span> <span>Bill Gates</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Strict reliance on numbers for decision making.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>Numbers are great. The main strategy to avoid the first two anti-patterns mentioned in this article (premature optimization and bikeshedding) was to profile or do A/B testing to get some measurements that can help you optimize or decide based on numbers instead of speculating. However, blind reliance on numbers can be dangerous. For example, numbers tend to outlive the models in which they were meaningful, or the models become outdated and no longer accurately represent reality. This can lead to poor decisions, especially if they are fully automated &#8213; see <a href="http://en.wikipedia.org/wiki/Automation_bias">automation bias</a>.</p>
<div>
<img alt='Pryzbylewski from the show "The Wire" teaching a classroom.' src="http://sahandsaba.com/images/pryzbylewski.jpg">
<p>Do you find yourself commiserating with Pryzbylewski from the HBO show The Wire, Season 4?</p>
</div>
<p>Another issue with reliance on numbers for <em>determining</em> (and not merely <em>informing</em>) decisions is that the measurement processes can be manipulated over time to achieve the desired numbers instead &#8213; see <a href="http://en.wikipedia.org/wiki/Observer-expectancy_effect">observer-expectancy effect</a>. Grade inflation is an example of this. The HBO show The Wire (which, by the way, if you haven't seen, you must!) does an excellent job of portraying this issue of reliance on numbers, by showing how the police department and later the education system have replaced meaningful goals with a game of numbers. Or if you prefer charts, the following one showing the distribution of scores on a test with a passing score of 30%, illustrates the point perfectly.</p>
<div>
<img alt="Score distribution of the 2013 high school exit exam in Poland with passing score of 30%." src="http://sahandsaba.com/images/poland-matura.png">
<p>Score distribution of the high school exit exam in Poland with passing score of 30%. <a href="http://www.cke.edu.pl/images/files/matura/informacje_o_wynikach/2013/2013_Matura.pdf">Source in Polish</a>, and the the <a href="http://www.reddit.com/r/dataisbeautiful/comments/27dx4q/distribution_of_results_of_the_matura_high_school/">Reddit post</a> that I first saw this in.</p>
</div>
</div>
<div>
<h3>How to avoid it</h3>
<p>Use measurements and numbers wisely, not blindly.</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>Using only lines of code, number of commits, etc. to judge the effectiveness of programmers. Measuring employee contribution by the numbers of hours they spend at their desks.</p>
</div>
<div>
<h3>The tricky part</h3>
<p>The larger the scale of operations, the higher the number of decisions that will need to be made, and this means automation and blind reliance on numbers for decisions begins to creep into the processes.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Use numbers to inform your decisions, not determine them.</p>
</div>
</div>
<div>
<h2>9&nbsp;&nbsp;&nbsp;Useless (Poltergeist) Classes</h2>
<blockquote>
<span>It seems that perfection is attained, not when there is nothing more to add, but when there is nothing more to take away.</span> <span>Antoine de Saint Exup&eacute;ry</span>
</blockquote>
<div>
<h3>What is it?</h3>
<p>Useless classes with no real responsibility of their own, often used to just invoke methods in another class or add an unneeded layer of abstraction.</p>
</div>
<div>
<h3>Why it's bad</h3>
<p>Poltergeist classes add complexity, extra code to maintain and test, and make the code less readable&mdash;the reader first needs to realize what the poltergeist does, which is often almost nothing, and then train herself to mentally replace uses of the poltergeist with the class that actually handles the responsibility.</p>
</div>
<div>
<h3>How to avoid it</h3>
<p>Don't write useless classes, or refactor to get rid of them. Jack Diederich has a great talk titled <a href="https://www.youtube.com/watch?v=o9pEzgHorH0">Stop Writing Classes</a> that is related to this anti-pattern.</p>
</div>
<div>
<h3>Examples and signs</h3>
<p>A couple of years ago, while working on my master's degree, I was a teaching assistant for a first-year Java programming course. For one of the labs, I was given the lab material which was to be on the topic of stacks and using linked lists to implement them. I was also given the reference "solution". This is the solution Java file I was given, almost verbatim (I removed the comments to save some space):</p>
<div>
<pre>
<span>import</span> <span>java.util.EmptyStackException</span><span>;</span>
<span>import</span> <span>java.util.LinkedList</span><span>;</span>

<span>public</span> <span>class</span> <span>LabStack</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>{</span>
    <span>private</span> <span>LinkedList</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>list</span><span>;</span>

    <span>public</span> <span>LabStack</span><span>()</span> <span>{</span>
        <span>list</span> <span>=</span> <span>new</span> <span>LinkedList</span><span>&lt;</span><span>T</span><span>&gt;();</span>
    <span>}</span>

    <span>public</span> <span>boolean</span> <span>empty</span><span>()</span> <span>{</span>
        <span>return</span> <span>list</span><span>.</span><span>isEmpty</span><span>();</span>
    <span>}</span>

    <span>public</span> <span>T</span> <span>peek</span><span>()</span> <span>throws</span> <span>EmptyStackException</span> <span>{</span>
        <span>if</span> <span>(</span><span>list</span><span>.</span><span>isEmpty</span><span>())</span> <span>{</span>
            <span>throw</span> <span>new</span> <span>EmptyStackException</span><span>();</span>
        <span>}</span>
        <span>return</span> <span>list</span><span>.</span><span>peek</span><span>();</span>
    <span>}</span>

    <span>public</span> <span>T</span> <span>pop</span><span>()</span> <span>throws</span> <span>EmptyStackException</span> <span>{</span>
        <span>if</span> <span>(</span><span>list</span><span>.</span><span>isEmpty</span><span>())</span> <span>{</span>
            <span>throw</span> <span>new</span> <span>EmptyStackException</span><span>();</span>
        <span>}</span>
        <span>return</span> <span>list</span><span>.</span><span>pop</span><span>();</span>
    <span>}</span>

    <span>public</span> <span>void</span> <span>push</span><span>(</span><span>T</span> <span>element</span><span>)</span> <span>{</span>
        <span>list</span><span>.</span><span>push</span><span>(</span><span>element</span><span>);</span>
    <span>}</span>

    <span>public</span> <span>int</span> <span>size</span><span>()</span> <span>{</span>
        <span>return</span> <span>list</span><span>.</span><span>size</span><span>();</span>
    <span>}</span>

    <span>public</span> <span>void</span> <span>makeEmpty</span><span>()</span> <span>{</span>
        <span>list</span><span>.</span><span>clear</span><span>();</span>
    <span>}</span>

    <span>public</span> <span>String</span> <span>toString</span><span>()</span> <span>{</span>
        <span>return</span> <span>list</span><span>.</span><span>toString</span><span>();</span>
    <span>}</span>
<span>}</span>
</pre>
</div>
<p>You can only imagine my confusion looking at the reference solution, trying to figure what the point of the <code>LabStack</code> class was, and what the students were supposed to learn from the utterly pointless exercise of writing it. In case it's not painfully obvious what's wrong with the class, it's that it does absolutely nothing! It simply passes calls through to the <code>LinkedList</code> object it instantiates. The class changes the names of a couple of methods (e.g. <code>makeEmpty</code> instead of the commonly used <code>clear</code>), which will only lead to user confusion. The error checking logic is completely unnecessary since the methods in <code>LinkedList</code> already do the same (but throw a different exception, <code>NoSuchElementException</code>, yet another possible source of confusion). To this day, I can't imagine what was going through the authors' minds when they came up with this lab material. Anytime you see classes that do anything similar to the above, reconsider whether they are really needed or not.</p>
<p>Update (May 23rd, 2015): There were interesting discussions over whether the <code>LabStack</code> class example above is a good example or not on Hacker News as well below in the comments. To clarify, I picked this class as a simple example for two reasons: firstly, in the context of teaching students about stacks, it is (almost) completely useless; and secondly, it adds unnecessary and duplicated code with the error-handling code that is already handled by <code>LinkedList</code>. I would agree that in other contexts, such classes can be useful but even in those cases, duplicating the error checking and throwing a semi-deprecated exception instead of the standard one and renaming methods to less-commonly-used names would be bad practice.</p>
</div>
<div>
<h3>The tricky part</h3>
<p>The advice here at first glance looks to be in direct contradiction of the advice in "Fear of Adding Classes". It's important to know when classes perform a valuable role and simplify the design, instead of uselessly increasing complexity with no added benefit.</p>
</div>
<div>
<h3>tl;dr</h3>
<p>Avoid classes with no real responsibility.</p>
</div>
</div>
<br>
<br>
<br>
<br>
Python<br>
<br>
via Math &cup; Code http://sahandsaba.com/<br>
<br>
May 9, 2015 at 10:25AM
